sos/report/plugins/lvm2.py

```diff
@@ -38,17 +38,21 @@ def do_lvmdump(self, metadata=False):
 
     def setup(self):
         # When running LVM2 comamnds:
-        # - use locking_type 0 (no locks) from lvm.conf: Turn locking
-        #   off by setting to 0 (dangerous: risks metadata corruption if
-        #   LVM2 commands get run concurrently).  This avoids the
-        #   possibility of hanging lvm commands when another process or
-        #   node holds a conflicting lock.
+        # - use nolocking if supported, else locking_type 0 (no locks)
+        #   from lvm.conf: Turn locking off by setting to 0 (dangerous:
+        #   risks metadata corruption if LVM2 commands get run
+        #   concurrently).  This avoids the possibility of hanging lvm
+        #   commands when another process or node holds a conflicting
+        #   lock.
         # - use metadata_read_only 1 (forbid on-disk changes). Although
         #   all LVM2 commands we use should be read-only, any LVM2
         #   command may attempt to recover on-disk data in some cases.
         #   This option prevents such changes, allowing safe use of
         #   locking_type=0.
-        lvm_opts = '--config="global{locking_type=0 metadata_read_only=1}"'
+        if '--nolocking' in self.exec_cmd('vgdisplay -h')['output']:
+            lvm_opts = '--config="global{metadata_read_only=1}" --nolocking'
+        else:
+            lvm_opts = '--config="global{locking_type=0 metadata_read_only=1}"'
```

Comment:
We have a somewhat recent addition in the API in the form of predicates that handles this directly.

It'd be something like:

```python
nolock = {'cmd': vgdisplay -h', 'output': '--nolocking'}
if self.test_predicate(self, pred=SoSPredicate(self, cmd_outputs=nolock):
    lvm_opts = 'foo'
else:
    lvm_opts = 'bar'
```

The `grub2` plugin has an example of this kind of predicate.

Reply:
Thanks! That seems to work (though the first `self` arg to `test_predicate` is unnecessary here and in `grub2.py`).

Reply:
You don't really need to use `test_predicate()` here since this is a known predicate for a specific condition - the only difference between `self.test_predicate(pred)` and `bool(pred)` is that `test_predicate()` will use the default command or plugin predicate instead of the `pred=` kwarg if they are set and it is not - that's probably not the behaviour you want here, even if it currently gives the same result.

The `test_predicate()` call has `self` as the first argument since it is a method of the `Plugin` class (in order to retrieve the stored command or plugin predicate).

Same comment applies to the use in `grub2`: it can simply be rewritten as:

```python
        if bool(SoSPredicate(self, cmd_outputs=co)):
            grub_cmd += ' --no-grubenv-update'
```

And likewise here - generally speaking plugins should not need to call `Plugin.test_predicate()` themselves - its main use is to consistently apply the predicate precedence rules when evaluating predicates in the `Plugin` helper methods (`add_copy_spec()`, `add_cmd_output()`, `add_string_as_file()`, `collect_cmd_output()`, and `exec_cmd()`: these need to properly evaluate any predicate passed in the method arguments as well as any stored command or plugin predicate. For this kind of one-off use where a single, specific predicate guards an action that's not required and it makes the code harder to follow.

Reply:
Great to know! That's much cleaner.